capa/features/extractors/dexfile.py

```diff
@@ -0,0 +1,418 @@
+            else:
+                source_file = ""
+
+            # Data
+            data_offset = clazz["class_data_off"]
+            if data_offset != 0:
+                data = self.dex.get_class_data(data_offset)
+            else:
+                data = None
+
+            classes[class_type] = DexAnalyzedClass(
+                offset=offset + index * 32,
+                class_type=class_type,
+                superclass_type=superclass_type,
+                interfaces=interfaces,
+                source_file=source_file,
+                data=data,
+            )
+        return classes
+
+    def _analyze_methods(self):
+        methods: List[DexAnalyzedMethod] = []
+        for method_id in self.method_ids:
+            proto = self.proto_ids[method_id["proto_idx"]]
+            parameters = []
+
+            param_off = proto["param_off"]
+            if param_off != 0:
+                size = struct.unpack("<L", self.dex.data[param_off : param_off + 4])[0]
+                for i in range(size):
+                    type_idx = struct.unpack("<H", self.dex.data[param_off + 4 + i * 2 : param_off + 6 + i * 2])[0]
+                    param_type = self.get_pretty_type(type_idx)
+                    parameters.append(param_type)
+
+            methods.append(
+                DexAnalyzedMethod(
+                    class_type=self.get_pretty_type(method_id["class_idx"]),
+                    name=self.get_string(method_id["name_idx"]),
+                    shorty_descriptor=self.get_string(proto["shorty_idx"]),
+                    return_type=self.get_pretty_type(proto["return_type_idx"]),
+                    parameters=parameters,
+                )
+            )
+
+        # Fill in the missing method data
+        for clazz in self.classes.values():
+            if clazz.data is None:
+                continue
+
+            for method_def in clazz.data["direct_methods"]:
+                diff = method_def["diff"]
+                methods[diff].access_flags = method_def["access_flags"]
+                methods[diff].code_offset = method_def["code_off"]
+
+            for method_def in clazz.data["virtual_methods"]:
+                diff = method_def["diff"]
+                methods[diff].access_flags = method_def["access_flags"]
+                methods[diff].code_offset = method_def["code_off"]
+
+        # Fill in the missing code offsets with fake data
+        offset = self.dex.header_data["method_ids_off"]
+        for index, method in enumerate(methods):
+            method.id_offset = offset + index * 8
+
+        return methods
+
+    def extract_file_features(self) -> Iterator[Tuple[Feature, Address]]:
+        yield Format(FORMAT_DEX), NO_ADDRESS
+
+        for i in range(len(self.strings)):
+            yield String(self.strings_utf8[i]), FileOffsetAddress(self.strings[i][0])
+
+        for method in self.methods:
+            yield FunctionName(method.qualified_name), DexMethodAddress(method.address)
```

Comment:
Do these include forward declared methods e.g. `Java.io.File::exists`?

Reply:
Yes, this includes all the methods. Otherwise you wouldn't be able to match `function-name: Java.io.File.exists`

Reply:
Gotcha' - capa's `function-name` feature has an overloaded meaning. `function-name` in the context of native code is a static library function that was identified using something like `FLIRT`. `function-name` in the context of managed code is a user-defined (not external) function.

We'll want to emit forward declared methods as `import` features at the file scope. For reference, here is a breakdown of how we treat `.NET` features:

- `namespace` e.g., `System.IO`
- `class` e.g., `System.IO.File`
- `api` and `import` e.g., `System.IO.File::Delete`
- `function-name` e.g., `HelloWorld::Main`

So for the forward declared method `Java.io.File::exists` we'd want to emit the following features at the file scope:
- `namespace(Java.io)`
- `class(Java.io.File)`
- `import(Java.io.File::exists)`

where `Java.io.File::exists` uses the format `<namespace>.<class>::<method>`.

Reply:
Good point. I'd also be in favor of following the Java syntax but the consistency is something to consider since we reuse all the underlying features.